text/0016-revise_transform_processing.md

```diff
@@ -66,41 +66,44 @@ Ideally, consumers of the return value of `download` have access to whatever the
 
-As a special function, `download` now has a unique signature that more closely matches it's intention. Download is intented to _download things_. It has no access to anything local, which is denoted by the empty function signature. Download should do network stuff, then return the results for further processing in `transform`. Transform should modify the dataset, so it's provided a dataset param. All of this will need to be thoroughly documented, and ideally made as a "tab completion" in the Qri frontend editor.
+`context` is designed to solve the issue of moving state around during a transform. All special functions will accept a `context` argument. The Qri environment will populate `context` with necessary transfrom state such as `transform.config` and `transform.secret` values. `context` also has an API for passing arbitrary user data via `context.set("key", Value)` and `context.get("key")`.
 
-Making the return value of `download` available at `qri.results.download` also helps clarify to the user that by the time `transform` is executing, the qri environment has already called download. This clarifies both what is doing the calling, and when it's been called. Adding to `results` instead of attaching directly to the `qri` object also reduces the chances of API namespace collision.
+_Note: Most special functions will not be able to reference reference state that another function procduces–even via context–because many special functions and will be called in parallel._
 
-A side effect of these changes, transform scripts now must define a `transform` function if they wish to have any effect on a dataset. Reducing a transform to this single requirement will make for a consistant point of entry that's easier for both humans and machines to reason about.
+As a special function, `download` now has a unique signature that more closely matches it's intention. Download is intented to _download things_. It has no access to the input dataset, which is denoted by the signature. Download should do network stuff, then return the results for further processing in `transform`. Transform should modify the dataset, so it's provided a dataset param. All of this will need to be thoroughly documented, (and ideally made as a "tab completion" in the Qri frontend editor), but this change helps convey intended use.
+
+Making the return value of `download` available at `context.download` also helps clarify to the user that by the time `transform` is executing, the qri environment has already called download. This clarifies both what is doing the calling, and when it's been called.
```

Comment:
Ah, I missed this part of the change before. Though I understand the objections to using "results", I am not as excited about `context.download` as `context.{something}.download` because it feels to me like part of the hierarchy is missing; "download" isn't directly owned by "context", it's owned by something that tracks the outputs of other functions calls, which is itself owned by context. From talking with @b5, I suggested `context.from.download`. But on the other hand, this isn't a huge deal, and it would be easier to fix this in the future (introduce something like `from` and deprecate the direct `context.download` method) rather than do the opposite.